sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala

```diff
@@ -882,6 +882,30 @@ class FilterPushdownSuite extends PlanTest {
     comparePlans(optimized, correctAnswer)
   }
 
+  test("union part 2 electric razor idk") {
+    val nonNullableArray = StructField("a", ArrayType(IntegerType, false))
+    val bField = StructField("b", IntegerType)
+    val testRelationNonNull = LocalRelation(nonNullableArray, bField)
+    val testRelationNull = LocalRelation($"c".array(IntegerType), $"d".int)
+
+    val nonNullArrayRef = AttributeReference("a", ArrayType(IntegerType, false))(
+      testRelationNonNull.output(0).exprId, List())
+
+
+    val originalQuery = Union(Seq(testRelationNonNull, testRelationNull))
+      .where(IsNotNull(nonNullArrayRef))
```

Comment:
I'm a bit confused about how such logical plans can be generated in the real world. Ideally the expressions in a query plan node can only references attributes from its children nodes. The `Filter` here should only references attributes from `Union.output`, but `nonNullArrayRef` is not from `Union.output`.

Reply:
Right so in the "real world" I think the way one would do it is with the data frame API and writing a filter referencing the original sub data frame but on the unions output. It's... a very weird corner case but one I've seen in production.

Reply:
ah, this reminds me of the rule `UpdateAttributeNullability` which does very similar things for the same reason. It seems we need to update the nullability of array/map elements and nested fields as well.

I think the dataframe col API is too powerful as it carries `AttributeReference`, and can lead to undefined behavior. e.g. `df1.union(df2).filter(df1("col").xxx)` will probably go wrong if `df1` and `df2` have different column data types and during the predicate evaluation bad things can happen as the actual value comes from Union will have a different data type than `df1("col")`.

That said, I think we should not blindly look at the expr ID and ignore the data type difference. We should only allow nullability different, and handle it in `UpdateAttributeNullability`.

Reply:
So the problem is for nested structures we can have a data type change since the nullability of the data type can be promoted as well.

exprID is unique right? What's the downside to depending on it?

Reply:
right, it doesn't hurt to look at expr id only here. It won't change the behavior.

After a second thought, `UpdateAttributeNullability` is also called in the optimizer, so it's safer to make changes to the filter pushdown rule.